conans/test/functional/compression/compression_test.py

```diff
@@ -0,0 +1,443 @@
+import abc
+import ctypes
+import os
+import random
+import six
+import stat
+import string
+import sys
+import tarfile
+import time
+import uuid
+import zipfile
+
+import unittest
+from nose.plugins.attrib import attr
+from parameterized import parameterized_class
+
+from conans.client.tools.files import chdir, load
+from conans.util.files import md5sum
+from conans.test.utils.tools import TestClient
+
+
+# more portable implementation, works on Windows 10 with developer mode (non-admin symbolic links)
+def create_symlink(src, dst):
+    # on Windows, special treatment is needed
+    if os.name == 'nt':
+        flags = 2  # SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE
+        if os.path.isdir(src):
+            flags |= 1  # SYMBOLIC_LINK_FLAG_DIRECTORY
+        CreateSymbolicLinkW = ctypes.windll.kernel32.CreateSymbolicLinkW
+        CreateSymbolicLinkW.argtypes = (ctypes.c_wchar_p, ctypes.c_wchar_p, ctypes.c_uint32)
+        CreateSymbolicLinkW.restype = ctypes.c_ubyte
+        res = CreateSymbolicLinkW(dst, src, flags)
+        if res == 0:
+            raise ctypes.WinError()
+    else:
+        os.symlink(src, dst)
+
+
+def is_link(path):
+    if os.name == 'nt':
+        FILE_ATTRIBUTE_REPARSE_POINT = 0x0400
+        INVALID_FILE_ATTRIBUTES = 0xFFFFFFFF
+        GetFileAttributesW = ctypes.windll.kernel32.GetFileAttributesW
+        GetFileAttributesW.argtypes = (ctypes.c_wchar_p,)
+        GetFileAttributesW.restype = ctypes.c_uint32
+
+        res = GetFileAttributesW(path)
+        if res == INVALID_FILE_ATTRIBUTES:
+            raise ctypes.WinError()
+        return FILE_ATTRIBUTE_REPARSE_POINT == (res & FILE_ATTRIBUTE_REPARSE_POINT)
+    return os.path.islink(path)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class Compressor:
+    @abc.abstractmethod
+    def compress(self, d):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def decompress(self, d):
+        raise NotImplementedError()
+
+
+class ZipCompressor(Compressor):
+    ext = ".zip"
+
+    def compress(self, d):
+        with zipfile.ZipFile(file='test' + self.ext, mode='w', compression=zipfile.ZIP_DEFLATED) as f:
+            for root, dirnames, filenames in os.walk(d):
+                for filename in filenames:
+                    filename = os.path.join(root, filename)
+                    archname = os.path.relpath(filename, d)
+                    if is_link(filename):
+                        target = os.readlink(filename)
+                        zi = zipfile.ZipInfo(archname)
+                        permissions = 0
+                        permissions |= 0xA000
```

Comment:
Is the first permissions = 0 redundant and this could be ``permissions = 0xA000``?
Could you please explain these permissions?

Reply:
yes, right now it's redundant. in general, some operation systems (e.g. macOS) allow to have different permissions for symlink itself and target. it's not cross-platform, and I don't test it here, but in general external attribute in Zip archive is combination of flags and permissions (same as stat **st_mode** field). I'll add links to the documentation for the clarification.

Reply:
added some comment to explain